File: Cpp/0886_possible_bipartition.cpp
// https://leetcode.cn/problems/possible-bipartition/

class Solution {
public:
    bool possibleBipartition(int n, vector<vector<int>>& dislikes) {
        // people[i] = 1: first group
        // people[i] = 2: second group
        vector<int> people(n, 0);
        unordered_map<int, vector<int>> table;
        for (vector<int> dis: dislikes) {
            table[dis[0]-1].push_back(dis[1]-1);
            table[dis[1]-1].push_back(dis[0]-1);
        }
        for (int i = 0; i < n; ++i) {
            if (people[i] == 0) {
                bool tmp = dfs(people, table, i, 1);
                if (!tmp) return false;
            }
        }
        return true;
    }

private:
    bool dfs(vector<int>& people, unordered_map<int, vector<int>>& table,
        int person, int color) {
            people[person] = color;
            for (int i: table[person]) {
                if (people[i] == color) return false;
                else if (people[i] == 0) {
                    bool tmp = dfs(people, table, i, 3-color);
                    if (!tmp) return false;
                }
            }
            return true;
        }
};

File: Python3/0886_possible_bipartition.py
# https://leetcode.cn/problems/possible-bipartition/

class Solution:
    def possibleBipartition(self, n, dislikes):
        """
        Inputs:
            n: int
            dislikes: list[list[int]]
        Outputs:
            res: bool
        """
        table = {i: [] for i in range(n)}
        for a, b in dislikes:
            table[a-1].append(b-1)
            table[b-1].append(a-1)
        # people[i] = 1: first group
        # people[i] = 2: second group
        people = [0 for i in range(n)]

        def dfs(person, group):
            people[person] = group
            for i in table[person]:
                if people[i] == group:
                    return False
                elif people[i] == 0:
                    tmp = dfs(i, 3-group)
                    if not tmp:
                        return False
            return True
        
        for i in range(n):
            if people[i] == 0:
                tmp = dfs(i, 1)
                if not tmp:
                    return False
        return True